### PR TITLE
Fix EditDepartmentRequestValidator and DepartmentUsersValidator

### PR DESCRIPTION
**Pull request by vorobevamv**
http://outsourcing.nat.tepkom.ru:20080/redmine/issues/1024

### CODE REVIEW

File: src/DepartmentService.Validation/DepartmentUsersValidator.cs [hunk 1]
```diff
@@ -1,10 +1,9 @@
...
 using System.Collections.Generic;
 using System.Linq;
-using System.Threading.Tasks;
 using FluentValidation;
+using LT.DigitalOffice.DepartmentService.Broker.Requests.Interfaces;
 using LT.DigitalOffice.DepartmentService.Validation.Interfaces;
-using LT.DigitalOffice.Kernel.BrokerSupport.Broker;
 using LT.DigitalOffice.Models.Broker.Common;
 using MassTransit;
 using Microsoft.Extensions.Logging;
```
Comment (MariyaSin) [inline src/DepartmentService.Validation/DepartmentUsersValidator.cs:9]: remove

File: src/DepartmentService.Validation/DepartmentUsersValidator.cs [hunk 2]
```diff
@@ -1,10 +1,9 @@
 ﻿using System;
 using System.Collections.Generic;
 using System.Linq;
-using System.Threading.Tasks;
 using FluentValidation;
+using LT.DigitalOffice.DepartmentService.Broker.Requests.Interfaces;
 using LT.DigitalOffice.DepartmentService.Validation.Interfaces;
-using LT.DigitalOffice.Kernel.BrokerSupport.Broker;
 using LT.DigitalOffice.Models.Broker.Common;
 using MassTransit;
```
[the remark this answers is not in the document]
Comment (vorobevamv) [reply src/DepartmentService.Validation/DepartmentUsersValidator.cs]: done

File: src/DepartmentService.Validation/DepartmentUsersValidator.cs (hunk 1 above)
Comment (vorobevamv) [reply src/DepartmentService.Validation/DepartmentUsersValidator.cs]: done

File: src/DepartmentService.Validation/CreateDepartmentRequestValidator.cs [hunk 3]
```diff
@@ -1,32 +1,20 @@
...
-using LT.DigitalOffice.Kernel.BrokerSupport.Broker;
-using LT.DigitalOffice.Models.Broker.Common;
-using MassTransit;
-using Microsoft.Extensions.Logging;
 
 namespace LT.DigitalOffice.DepartmentService.Validation.Department
 {
   public class CreateDepartmentRequestValidator : AbstractValidator<CreateDepartmentRequest>, ICreateDepartmentRequestValidator
   {
     private readonly IDepartmentRepository _repository;
```
Comment (MariyaSin) [inline src/DepartmentService.Validation/CreateDepartmentRequestValidator.cs:11]: remove

File: src/DepartmentService.Validation/CreateDepartmentRequestValidator.cs [hunk 4]
```diff
@@ -1,32 +1,20 @@
...
-    private readonly IRequestClient<ICheckUsersExistence> _rcCheckUsersExistence;
-    private readonly ILogger<CreateDepartmentRequestValidator> _logger;
 
     public CreateDepartmentRequestValidator(
       IDepartmentRepository repository,
-      IRequestClient<ICheckUsersExistence> rcCheckUsersExistence,
-      ILogger<CreateDepartmentRequestValidator> logger)
+      IDepartmentUsersValidator departmentusersvalidator)
     {
       _repository = repository;
```
Comment (MariyaSin) [inline src/DepartmentService.Validation/CreateDepartmentRequestValidator.cs:17]: remove

File: src/DepartmentService.Validation/CreateDepartmentRequestValidator.cs [hunk 5]
```diff
@@ -50,53 +38,16 @@ public class CreateDepartmentRequestValidator : AbstractValidator<CreateDepartme
             RuleForEach(department => department.Users)
```
Comment (MariyaSin) [inline src/DepartmentService.Validation/CreateDepartmentRequestValidator.cs:38]: do you need cascade here?

File: src/DepartmentService.Validation/CreateDepartmentRequestValidator.cs [hunk 6]
```diff
@@ -50,53 +38,16 @@ public class CreateDepartmentRequestValidator : AbstractValidator<CreateDepartme
...
               }))
-          .Must(d => d.Select(du => du.UserId).Distinct().Count() == d.Count())
-          .WithMessage("User cannot be added to the deaprtment twice.")
           .Must(d => d.Where(du => du.Role == Models.Dto.Enums.DepartmentUserRole.Director).Count() < 2)
           .WithMessage("Only one user can be the department director")
-          .MustAsync(async (d, _) => await CheckUsersExistenceAsync(d.Select(du => du.UserId).ToList()))
-          .WithMessage("Some users does not exist.");
+          .ChildRules(d =>
+             RuleFor(d => d.Users.Select(user => user.UserId).ToList())
+             .SetValidator(departmentusersvalidator));
```
Comment (MariyaSin) [inline src/DepartmentService.Validation/CreateDepartmentRequestValidator.cs:48]: add tab

File: src/DepartmentService.Validation/CreateDepartmentRequestValidator.cs (hunk 3 above)
Comment (vorobevamv) [reply src/DepartmentService.Validation/CreateDepartmentRequestValidator.cs:11]: deleted

File: src/DepartmentService.Validation/CreateDepartmentRequestValidator.cs (hunk 4 above)
Comment (vorobevamv) [reply src/DepartmentService.Validation/CreateDepartmentRequestValidator.cs:17]: deleted

File: src/DepartmentService.Validation/CreateDepartmentRequestValidator.cs (hunk 5 above)
Comment (vorobevamv) [reply src/DepartmentService.Validation/CreateDepartmentRequestValidator.cs:38]: removed

File: src/DepartmentService.Validation/CreateDepartmentRequestValidator.cs (hunk 6 above)
Comment (vorobevamv) [reply src/DepartmentService.Validation/CreateDepartmentRequestValidator.cs:48]: done

File: src/DepartmentService.Validation/CreateDepartmentRequestValidator.cs [hunk 7]
```diff
@@ -42,61 +28,25 @@ public class CreateDepartmentRequestValidator : AbstractValidator<CreateDepartme
...
-        _logger.LogError(
-          exc,
-          "Can not check users existence Ids: {UsersIds}.",
-          string.Join(", ", usersIds));
-      }
-
-      return false;
+      RuleFor(request => request.Users)
+        .NotNull()
+        .WithMessage("Users should not be empty.")
```
Comment (MariyaSin) [inline src/DepartmentService.Validation/CreateDepartmentRequestValidator.cs:33]: "Users should not be null."

File: src/DepartmentService.Validation/CreateDepartmentRequestValidator.cs [hunk 8]
```diff
@@ -42,61 +28,25 @@ public class CreateDepartmentRequestValidator : AbstractValidator<CreateDepartme
...
+            dus.When(dus => dus.Any(), () =>
+            {
+                dus.RuleForEach(dus => dus)
+                  .ChildRules(du =>
+                      du.RuleFor(u => u.Role)
+                        .IsInEnum().WithMessage("Wrong type of user role."));
+
+                dus.RuleFor(dus => dus)
+                  .Must(dus => dus.Where(u => u.Role == Models.Dto.Enums.DepartmentUserRole.Director).Count() < 2)
+                  .WithMessage("Only one user can be the department director")
```
[the remark this answers is not in the document]
Comment (vorobevamv) [reply src/DepartmentService.Validation/CreateDepartmentRequestValidator.cs:44]: done

**Review by VLiamin — APPROVED** [review]

**Review by a5x55555 — APPROVED** [review]

**Review by AnnPostnikova — APPROVED** [review]

File: src/DepartmentService.Business/Department/CreateDepartmentCommand.cs [hunk 9]
```diff
@@ -64,7 +64,7 @@ public class CreateDepartmentCommand : ICreateDepartmentCommand
 
       #region Deactivated previous department user records
 
-      if (request.Users != null)
+      if (request.Users.Any())
```
Comment (ndoriomedov) [inline src/DepartmentService.Business/Department/CreateDepartmentCommand.cs:67]: request.Users?.Any() maybe?

**Review by ndoriomedov — APPROVED** [review]

File: src/DepartmentService.Business/Department/CreateDepartmentCommand.cs (hunk 9 above)
Comment (vorobevamv) [reply src/DepartmentService.Business/Department/CreateDepartmentCommand.cs:67]: VS don't like this variation. There appear a problem with converting "bool?" to "bool".
So, as I understand, it's possible to stay variation with "request.Users.Any()" or change all if-construction.

Comment (vorobevamv) [reply src/DepartmentService.Business/Department/CreateDepartmentCommand.cs:67]: Check for null is in validation. So, it can't be null here.

**Pull request merged**